### PR TITLE
Support for KHR_materials_ior

### LINKED ISSUE #3652: Support glTF extension: KHR_materials_ior
https://github.com/KhronosGroup/glTF/blob/main/extensions/2.0/Khronos/KHR_materials_ior/README.md

The dielectric BRDF of the metallic-roughness material in glTF uses a fixed value of 1.5 for the index of refraction. This is a good fit for many plastics and glass, but not for other materials like water or asphalt, sapphire or diamond. This extensions allows users to set the index of refraction to a certain value.

### PR DESCRIPTION
**Pull request by GSterbrant**
### Description

This PR adds support for controlling index of refraction (IOR) from the frontend of the shader chunks, adds support for the KHR_materials_ior extension, as well as precalculates the F0 specular factor with is dependent on IOR.

Fixes #3652 

This image illustrates, in Y diffuse color from black to red, in X refraction index from 1 to 2.
image

This illustrates the visual change on the specular when IOR is changed, as that impacts the f0 specular component. As in the previous image, in Y diffuse color ranges from black to red, in X refraction index from 1 to 2.
image

### CODE REVIEW

File: src/scene/materials/standard-material.js [hunk 1]
```diff
@@ -220,6 +220,7 @@ let _params = new Set();
  * indices of refraction, the one around the object and the one of its own surface. In most
  * situations outer medium is air, so outerIor will be approximately 1. Then you only need to do
  * (1.0 / surfaceIor).
+ * @property {number} f0 Defines the f0 weight as calculated using ((ior - 1) / (ior + 1))^2
```
Comment (willeastcott) [inline src/scene/materials/standard-material.js:223]: Just wondering - does this really need to be public API? Would we expect somebody to set this in code considering it seems to be directly derived from the `ior (refractionIndex)` value?

**Review by willeastcott — APPROVED** [review]
Approving with a minor comment. This PR is a beautiful thing! 🌷

**Comment by mvaligursky** [comment]
Would you have some screenshot demonstrating what it can be used for?

**Comment by GSterbrant** [comment]
Added two screenshots showing both the refraction and how the specular effects change with the refraction index.

File: src/scene/materials/standard-material-options-builder.js [hunk 2]
```diff
@@ -162,6 +162,7 @@ class StandardMaterialOptionsBuilder {
         options.cubeMapProjection = stdMat.cubeMapProjection;
         options.customFragmentShader = stdMat.customFragmentShader;
         options.refraction = !!stdMat.refraction;
+        options.refractionIndexTint = (stdMat.refractionIndex !== 1.5) ? 1 : 0;
```
[the remark this answers is not in the document]
Comment (willeastcott) [reply src/scene/materials/standard-material-options-builder.js:165]: Aha, got it, OK.

**Review by mvaligursky — APPROVED** [review]
looks really good to me.

File: src/graphics/program-lib/programs/standard.js [hunk 3]
```diff
@@ -305,13 +305,18 @@ const standard = {
             code.append(this._addMap("diffuse", "diffusePS", options, litShader.chunks));
             func.append("getAlbedo();");
 
+            if (options.refraction) {
+                decl.append("float dIor;");
+                code.append(this._addMap("refractionIndex", "iorPS", options, litShader.chunks));
```
Comment (slimbuck) [inline src/graphics/program-lib/programs/standard.js:310]: This function assumes a full map - so you need to add all the iorPS members to material - iorPSMap, iorPSUv, iorPSTiling, iorPSOffset etc and then handle them in the shader.

Comment (willeastcott) [reply src/graphics/program-lib/programs/standard.js:310]: I think you mean `refractionIndexMap`, `refractionMapUv`, etc, right?

Comment (slimbuck) [reply src/graphics/program-lib/programs/standard.js:310]: Yes!

Though I see from https://github.com/KhronosGroup/glTF/blob/main/extensions/2.0/Khronos/KHR_materials_ior/README.md that refraction index is a per-material constant (not texture or vertex driven).

@GSterbrant is it even necessary to have a customisation chunk for this? Seems like ior is literally just a float per material. User can set that on javascript - no need for customising chunks etc.

**Pull request merged**